src/lib/Libifl/pbsD_deljoblist.c

```diff
@@ -46,160 +46,554 @@
+		return NULL;
+	}
+
+	/* Broadcast the jids to all servers */
+	for (i = 0; i < num_svrs; i++) {
+		if (send_deljob(conns[i]->sd, function, jobids, numids, extend))
+			skip_list[i] = 1;
+	}
+	for (i = 0; i < num_svrs; i++) {
+		if (!skip_list[i]) {
+			struct batch_deljob_status *ret = NULL;
+
+			ret = recv_deljob(conns[i]->sd);
+			svr_retlist[i] = ret;
+		}
+	}
+
+	/* The list of ids we return has to be the intersection of the lists from all servers
+	 * Use the first server's list as the intersection to be returned
+	 */
+	for (iter_list = svr_retlist[0]; iter_list != NULL; prev = iter_list, iter_list = next) {
+		int job_count = 1;
+		struct batch_deljob_status *iter_list2;
+		next = iter_list->next;
+		int error_code = PBSE_UNKJOBID;
+
+		/* Make sure that this jid is in ALL of the other lists as well */
+		for (i = 1; i < num_svrs; i++) {
+			for (iter_list2 = svr_retlist[i]; iter_list2 != NULL; iter_list2 = iter_list2->next) {
+				if (strcmp(iter_list2->name, iter_list->name) == 0) {
+					/* Check for error code other than PBSE_UNKJOBID */
+					if (iter_list2->code != PBSE_UNKJOBID)
+						error_code = iter_list2->code;
+					job_count++;
+					break;
+				}
+			}
+		}
+
+		if (job_count != num_svrs) {	/* Job was found (and deleted) on at least 1 server */
+			/* Remove the job from the list to return */
+			if (prev != NULL)
+				prev->next = next;
+			else
+				svr_retlist[0] = next;
+
+			iter_list->next = NULL;
+			pbs_delstatfree(iter_list);
+			iter_list = prev;
+		} else
+			iter_list->code = error_code;
+	}
+
+	retlist = svr_retlist[0];
+	for (i = 1; i < num_svrs; i++) {	/* Free the lists returned by all other servers */
+		pbs_delstatfree(svr_retlist[i]);
+	}
+	free(svr_retlist);
+	free(skip_list);
+
+	return retlist;
+}
+
+/**
+ * @brief	Handle deleting a list of jobs in batches
  *
- * @return      int
- * @retval      DIS_SUCCESS(0)  success
- * @retval      error code      error
+ * @param[in] c - communication handle
+ * @param[in] function - req type
+ * @param[in] jobids - the job id list
+ * @param[in] numjids - the count of job ids
+ * @param[in] extend - extend string for req
+ *
+ * @return     struct batch_deljob_status *
+ * @retval     list of jobs which couldn't be deleted
  *
  */
-int
-PBSD_deljoblist_put(int c, int function, char **jobids, int numofjobs, char *extend, int prot, char **msgid)
+static struct batch_deljob_status *
+PBSD_deljoblist(int c, int function, char **jobids, int numjids, char *extend)
 {
-	int rc;
+	int i;
+	svr_jobid_list_t *svr_itr = NULL;
+	struct batch_deljob_status *retlist = NULL;
+	svr_jobid_list_t *svr_jobid_list_hd = NULL;
+	char **bcastjids = NULL;
+	int bcastjidssize = 0;
+	int nbcastids = 0;
+	int fd_virtual = is_fd_virtual(c);
 
-	if (prot == PROT_TCP) {
-		DIS_tcp_funcs();
-	} else {
-		if ((rc = is_compose_cmd(c, IS_CMD, msgid)) != DIS_SUCCESS)
-			return rc;
-	}
+	if (fd_virtual) {	/* Virtual fd means we can talk to more than 1 server instance */
```

Comment:
Not necessarily according to your definition of is_fd_virtual(). There can be just one instance or n instances based on the cluster configuration.
If it is just one, you might still want to take the optimal path of a single server.

Reply:
ok, ive renamed the function to multi_svr_op() and am checking more than just get_conn_svr_instances() == NULL, let me know what you think

Reply:
Although what you are going to perform is an operation, the check itself is not. Would it be better to call it multi_svr(fd) without the _op part?

Reply:
We could be in multi-server setup but still the function will return False if fd is instance fd. So, I think the _op makes it a bit more clear that we are talking about single server operation vs multi-server operation, regardless of the setup. So, I'd prefer keeping it